File: googlewebtranslatortest/src/main/java/com/firemaples/googlewebtranslatortest/MainActivity.java
package com.firemaples.googlewebtranslatortest;

import android.content.DialogInterface;
import android.os.Bundle;
import android.support.v7.app.AlertDialog;
import android.support.v7.app.AppCompatActivity;
import android.util.Log;
import android.view.Menu;
import android.view.MenuItem;
import android.view.ViewGroup;
import android.widget.EditText;

import com.firemaples.googlewebtranslator.GoogleTranslator;
import com.firemaples.googlewebtranslator.GoogleWebTranslator;
import com.firemaples.googlewebtranslator.Language;
import com.firemaples.googlewebtranslator.TranslatedResult;

public class MainActivity extends AppCompatActivity {
    private static final String TAG = MainActivity.class.getSimpleName();

//    private GoogleWebTranslator translator;

    private GoogleTranslator googleTranslator;

    private GoogleWebTranslator.OnTranslationCallback onTranslationCallback = new GoogleWebTranslator.OnTranslationCallback() {

        @Override
        public void onInitialized(GoogleWebTranslator translator) {
            Log.i(TAG, "Translator initialized");
        }

        @Override
        public void onInitializationFailed(GoogleWebTranslator translator, int errorCode, String desc) {
            Log.e(TAG, "Translator initialization failed");
        }

        @Override
        public void onTranslationSuccess(GoogleWebTranslator translator, TranslatedResult result) {
            Log.i(TAG, "Translation success: " + result.text);
        }

        @Override
        public void onTranslationFailed(GoogleWebTranslator translator, Throwable throwable) {
            Log.e(TAG, "Translation failed: " + throwable);
        }
    };

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_main);

//        translator = GoogleWebTranslator.init(this);
        googleTranslator = new GoogleTranslator(this);

//        translator.addOnTranslationCallback(onTranslationCallback);

        setViews();
    }

    private void setViews() {
        ViewGroup rootView = findViewById(R.id.view_root);
//        rootView.addView(translator.getNonParentWebView(),
//                new ViewGroup.LayoutParams(
//                        ViewGroup.LayoutParams.MATCH_PARENT,
//                        ViewGroup.LayoutParams.MATCH_PARENT));

        googleTranslator.setup(rootView);

//        translator.setTargetLanguage(Language.Chinese_Traditional);
    }

    @Override
    protected void onDestroy() {
        super.onDestroy();
//        translator.removeOnTranslationCallback(onTranslationCallback);
    }

    @Override
    public boolean onCreateOptionsMenu(Menu menu) {
        getMenuInflater().inflate(R.menu.menu_main, menu);
        return true;
    }

    @Override
    public boolean onOptionsItemSelected(MenuItem item) {
        int id = item.getItemId();
        if (id == R.id.action_translate) {
            AlertDialog.Builder ab = new AlertDialog.Builder(this);
            final EditText et = new EditText(this);
            ab.setView(et);
            ab.setPositiveButton(android.R.string.ok, new DialogInterface.OnClickListener() {
                @Override
                public void onClick(DialogInterface dialogInterface, int i) {
                    String text = et.getText().toString();
//                    translator.translate(text);
                    googleTranslator.translate(text, Language.Chinese_Traditional.getLangCode());
                }
            });
            ab.setNegativeButton(android.R.string.cancel, null);
            ab.show();
            return true;
        }
        return super.onOptionsItemSelected(item);
    }
}

File: testapp/src/main/java/com/example/testapp/MainActivity.kt
package com.example.testapp

import android.os.Bundle
import androidx.appcompat.app.AppCompatActivity
import android.util.Log
import android.view.View
import android.view.ViewGroup
import android.webkit.WebView
import android.widget.*
import com.firemaples.googlewebtranslator.GoogleWebTranslator
import com.firemaples.googlewebtranslator.Language
import com.firemaples.googlewebtranslator.Tools
import com.firemaples.googlewebtranslator.TranslatedResult

class MainActivity : AppCompatActivity() {
    private val tag = MainActivity::class.java.simpleName
    private val translator by lazy { GoogleWebTranslator(this) }
    private val translatorNew by lazy {
        com.firemaples.googlewebtranslator.translator.GoogleWebTranslator(this).apply {
            onResult = { text, error ->
                tvResult.text = text ?: error
            }
        }
    }

    lateinit var tvResult: TextView

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_main)

        Tools.DEBUG = true

        WebView.setWebContentsDebuggingEnabled(true)

        val spLang = findViewById<Spinner>(R.id.sp_lang)
        val btSubmit = findViewById<Button>(R.id.bt_submit)
        val wrapWebView = findViewById<ViewGroup>(R.id.wrap_webView)
        val etText = findViewById<EditText>(R.id.et_text)
        tvResult = findViewById(R.id.tv_result)

        spLang.adapter = ArrayAdapter(this,
            android.R.layout.simple_spinner_dropdown_item, Language.values()
                .map { it.name })
//
//        translator.setup(wrapWebView)
        translatorNew.setup(wrapWebView)

        btSubmit.setOnClickListener {
            val text = etText.text
            if (!text.isNullOrBlank()) {
                translatorNew.translate(
                    text = text.toString(),
                    targetLang = Language.values()[spLang.selectedItemPosition].langCode
                )
//                translator.translate(_text = text.toString(),
//                    targetLang = Language.values()[spLang.selectedItemPosition].langCode,
//                    callback = object : GoogleWebTranslator.OnTranslationCallback {
//                        override fun onStart() {
//                            Log.d(tag, "onStart()")
//                        }
//
//                        override fun onTranslated(result: TranslatedResult) {
//                            Log.d(tag, "onTranslated(): ${result.text}")
//
//                            tvResult.text = result.text
//                        }
//
//                        override fun onTranslationFailed(errorMsg: String) {
//                            Log.e(tag, "onTranslationFailed: $errorMsg")
//
//                            tvResult.text = errorMsg
//                        }
//                    })
            }
        }
    }
}
